crates/matrix-sdk-ui/src/timeline/mod.rs

```diff
@@ -830,3 +835,157 @@ where
+    #[must_use]
+    pub fn add_item(mut self, item: GalleryItemInfo) -> Self {
+        self.items.push(item);
+        self
+    }
+
+    /// Set the optional caption.
+    ///
+    /// # Arguments
+    ///
+    /// * `caption` - The optional caption.
+    pub fn caption(mut self, caption: Option<String>) -> Self {
+        self.caption = caption;
+        self
+    }
+
+    /// Set the optional formatted caption.
+    ///
+    /// # Arguments
+    ///
+    /// * `formatted_caption` - The optional formatted caption.
+    pub fn formatted_caption(mut self, formatted_caption: Option<FormattedBody>) -> Self {
+        self.formatted_caption = formatted_caption;
+        self
+    }
+
+    /// Set the mentions of the message.
+    ///
+    /// # Arguments
+    ///
+    /// * `mentions` - The mentions of the message.
+    pub fn mentions(mut self, mentions: Option<Mentions>) -> Self {
+        self.mentions = mentions;
+        self
+    }
+
+    /// Set the reply information of the message.
+    ///
+    /// # Arguments
+    ///
+    /// * `reply` - The reply information of the message.
+    pub fn reply(mut self, reply: Option<Reply>) -> Self {
+        self.reply = reply;
+        self
+    }
+
+    /// Returns the number of media items in the gallery.
+    pub fn len(&self) -> usize {
+        self.items.len()
+    }
+
+    /// Checks whether the gallery contains any media items or not.
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
+}
+
+#[cfg(feature = "unstable-msc4274")]
+impl TryFrom<GalleryConfig> for matrix_sdk::attachment::GalleryConfig {
+    type Error = Error;
+
+    fn try_from(value: GalleryConfig) -> std::result::Result<Self, Self::Error> {
+        let mut config = matrix_sdk::attachment::GalleryConfig::new();
+
+        if let Some(txn_id) = value.txn_id {
+            config = config.txn_id(txn_id);
+        }
+
+        for item in value.items {
+            config = config.add_item(item.try_into()?);
+        }
+
+        config = config.caption(value.caption);
+        config = config.formatted_caption(value.formatted_caption);
+        config = config.mentions(value.mentions);
+        config = config.reply(value.reply);
+
+        Ok(config)
+    }
+}
+
+#[cfg(feature = "unstable-msc4274")]
+#[derive(Debug)]
+/// Metadata for a gallery item
+pub struct GalleryItemInfo {
+    /// The attachment source.
+    pub source: AttachmentSource,
+    /// The mime type.
+    pub content_type: Mime,
+    /// The attachment info.
+    pub attachment_info: AttachmentInfo,
+    /// The caption.
+    pub caption: Option<String>,
+    /// The formatted caption.
+    pub formatted_caption: Option<FormattedBody>,
+    /// The thumbnail.
+    pub thumbnail: Option<Thumbnail>,
+}
+
+#[cfg(feature = "unstable-msc4274")]
+impl TryFrom<GalleryItemInfo> for matrix_sdk::attachment::GalleryItemInfo {
+    type Error = Error;
+
+    fn try_from(value: GalleryItemInfo) -> std::result::Result<Self, Self::Error> {
```

Comment:
nit: do we have to precise `std::result::Result` here? If so, it might mean that the `Result` type we're defining crate/module wide could use two generic parameters, with the second one defaulting to the error type which might be hardcoded right now.

Reply:
No, actually not. Sorry, I think auto-completion in my IDE might have inserted this and I hadn't noticed.